lib/xwax/pitch.h

```diff
@@ -1,71 +1,213 @@
+    /* state */
+    double x;      /* position (relative; we subtract dx each step) */
+    double v;      /* velocity (pitch) */
+
+    /* covariance P (2x2, symmetric) */
+    double Pxx, Pxv, Pvv;
+
+    /* Residual treshold for the mode switch */
+    double mode_threshold;
+
+    /* Stable and reactive parameters for the mode switch */
+    double q_stable, r_stable;
+    double q_reactive, r_reactive;
 };
 
-/* Prepare the filter for observations every dt seconds */
+/*
+ * Initialize with dt and noise params.
+ * q: acceleration noise spectral density (tune up if motion is more jittery)
+ * r: variance of dx measurement (tune up if observations are noisier)
+ */
 
-static inline void pitch_init(struct pitch *p, double dt)
+static inline void pitch_init(struct pitch *p, double dt, double q_stable, double r_stable,
+        double q_reactive, double r_reactive, double mode_threshold)
 {
+    if (!p) {
+        errno = EINVAL;
+        perror(__func__);
+        return;
+    }
+
     p->dt = dt;
+
     p->x = 0.0;
     p->v = 0.0;
-}
 
-/* Input an observation to the filter; in the last dt seconds the
- * position has moved by dx.
- *
- * Because the vinyl uses timestamps, the values for dx are discrete
- * rather than smooth. */
+    /* Start with large uncertainty so the filter trusts early measurements */
+    p->Pxx = 1e6;
+    p->Pxv = 0.0;
+    p->Pvv = 1e6;
 
-static inline void pitch_dt_observation(struct pitch *p, double dx)
-{
-    double predicted_x, predicted_v, residual_x;
+    /* Concluded experimentally */
+    p->mode_threshold = mode_threshold;
+
+    /* Q and R for stable playback */
+    p->q_stable = q_stable;
+    p->r_stable = r_stable;
+
+    /* Q and R for scratching */
+    p->q_reactive = q_reactive;
+    p->r_reactive = r_reactive;
+
+    /* Use move semantics for efficiency */
+    p->q  = &p->q_reactive;
+    p->r  = &p->r_reactive;
+}
 
-    predicted_x = p->x + p->v * p->dt;
-    predicted_v = p->v;
+/*
+ * Optionally retune noise without resetting state
+ */
 
-    residual_x = dx - predicted_x;
+static inline void pitch_retune_noise(struct pitch *p, double *q, double *r)
+{
+    if (!p || !q || !r) {
+        errno = EINVAL;
+        perror(__func__);
+        return;
+    }
+
+    p->q = q;
+    p->r = r;
+}
 
-    p->x = predicted_x + residual_x * ALPHA;
-    p->v = predicted_v + residual_x * BETA / p->dt;
+/*
+ * Feed one observation: in the last dt seconds, position moved by dx
+ */
 
-    p->x -= dx; /* relative to previous */
+static inline void pitch_dt_observation(struct pitch *p, double dx)
+{
+    if (!p) {
+        errno = EINVAL;
+        perror(__func__);
+        return;
+    }
+
+    const double dt = p->dt;
+
+    /*
+     * ==== Predict step ====
+     * State transition F = [1 dt; 0 1]
+     * Process noise for constant-velocity model with white accel (G*q*G^T):
+     *   Q = q * [dt^3/3  dt^2/2;
+     *            dt^2/2  dt]
+     */
+
+    const double F11 = 1.0, F12 = dt;
```

Comment:
Is better readable IMHO 
```suggestion
    const double F11 = 1.0;
    const double F12 = dt;
```